### PR TITLE
Add zip response

### PR DESCRIPTION
**Pull request by jakobsack**
This pull request allows apps to return a zip file created on the fly.

implements #7589

### CODE REVIEW

File: lib/public/AppFramework/Http/ZipResponse.php [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+	private $request;
+
+	public function __construct(string $name = 'output', IRequest $request) {
+		$this->name = $name;
+		$this->request = $request;
+	}
+
+	public function addResource($r, string $internalName, int $size, int $time = -1) {
+		if (!\is_resource($r)) {
+			return;
```
Comment (ChristophWurst) [inline lib/public/AppFramework/Http/ZipResponse.php]: I wonder if ignoring this invalid argument silently is the best way to handle this. Maybe we should just fail with an InvalidArgumentExeption or similar. What do you think, @rullzer?

File: lib/public/AppFramework/Http/ZipResponse.php [hunk 2]
```diff
@@ -0,0 +1,75 @@
+<?php
+/**
+ * @copyright Copyright (c) 2016, ownCloud, Inc.
```
Comment (ChristophWurst) [inline lib/public/AppFramework/Http/ZipResponse.php]: :see_no_evil:

File: lib/public/AppFramework/Http/ZipResponse.php [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+ *
+ * @author Jakob Sack <mail@jakobsack.de>
+ * @author Roeland Jago Douma <roeland@famdouma.nl>
```
Comment (ChristophWurst) [inline lib/public/AppFramework/Http/ZipResponse.php]: really?

**Review by ChristophWurst — CHANGES_REQUESTED** [review]
Generally, this looks good. Thanks!

I'd love to see tests, though, if possible.

File: lib/public/AppFramework/Http/ZipResponse.php (hunk 3 above)
Comment (rullzer) [reply lib/public/AppFramework/Http/ZipResponse.php]: Yes he took it from an old PR of mine ;)

File: lib/public/AppFramework/Http/ZipResponse.php (hunk 1 above)
Comment (rullzer) [reply lib/public/AppFramework/Http/ZipResponse.php]: Agreed done

File: lib/public/AppFramework/Http/ZipResponse.php (hunk 2 above)
Comment (rullzer) [reply lib/public/AppFramework/Http/ZipResponse.php]: fixed

**Comment by rullzer** [comment]
I'm not sure yet how to test this as the zipstreamer is quite aggressive and just outputs.

We could of course wrap it in another stream. Thinking about it that probably is best for the abstraction.

**Comment by MorrisJobke** [comment]
@rullzer said: "we should abstract it with a stream wrapper"

**Review by rullzer — APPROVED** [review]

**Review by MorrisJobke — APPROVED** [review]

**Comment by jakobsack** [comment]
Hi all, thanks for merging this while I was on vacation :-)

**Pull request merged**